File: Java/src/main/java/QuestionBank/Solution001_100/Solution18.java
package QuestionBank.Solution001_100;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * @author windows
 */
public class Solution18 {

    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        ArrayList<List<Integer>> lists = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1]){
                continue;
            }
            for (int j = i + 1; j < nums.length; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]){
                    continue;
                }
                long sum = nums[i] + nums[j];
                int left = j + 1;
                int right = nums.length - 1;
                while(left < right){
                    if (nums[left] + nums[right] + sum < target){
                        left++;
                    }else if (nums[left] + nums[right] + sum > target){
                        right--;
                    }else{
                        lists.add(Stream.of(nums[i], nums[j], nums[left], nums[right]).collect(Collectors.toList()));
                        left++;
                        while (left < right && nums[left] == nums[left - 1]){
                            left++;
                        }
                        right--;
                        while (right > left && nums[right] == nums[right + 1]){
                            right--;
                        }
                    }
                }
            }
        }
        return lists;
    }


    public static void main(String[] args) {
        Solution18 solution18 = new Solution18();
        System.out.println(solution18.fourSum(new int[]{-2,-1,-1,1,1,2,2}, 0));
    }
}

File: C++/src/Solution1_100/Solution18.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution18.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <algorithm>

//vector<vector<int>> Solution::fourSum(vector<int> &nums, int target) {
//    sort(nums.begin(), nums.end());
//    int n = nums.size();
//    vector<vector<int>> res;
//    for (int i = 0; i < n - 3; ++i) {
//        if (i > 0 && nums[i] == nums[i -1]){
//            continue;
//        }
//        for (int j = i + 1; j < n - 2; ++j) {
//            if (j > i + 1 && nums[j] == nums[j - 1]){
//                continue;
//            }
//            int k = j + 1, l = n - 1;
//            while (k < l) {
//                long num = (long)nums[i] + nums[j] + nums[k] + nums[l];
//                if (num == target) {
//                    res.push_back({nums[i], nums[j], nums[k], nums[l]});
//                    do k++; while (k < l && nums[k] == nums[k - 1]);
//                    do l--; while (k < l && nums[l] == nums[l + 1]);
//                } else if (num < target) {
//                    k++;
//                }else{
//                    l--;
//                }
//            }
//        }
//    }
//    return res;
//}





vector<vector<int>> Solution::fourSum(vector<int> &nums, int target) {
    int n = nums.size();
    sort(nums.begin(), nums.end());
    vector<vector<int>> res;
    for (int i = 0; i < n - 3; ++i) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;
        if ((long)nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target){
            break;
        }
        if ((long)nums[i] + nums[n -1] + nums[n -2] + nums[n -3] < target){
            continue;
        }
        for (int j = i + 1; j < n - 2; ++j) {
            if (j > i + 1 && nums[j] == nums[j - 1]) continue;
            int k = j + 1, l = n - 1;
            while (k < l) {
                long num = (long) nums[i] + nums[j] + nums[k] + nums[l];
                if (target == num) {
                    res.push_back({nums[i], nums[j], nums[k], nums[l]});
                    do k++; while (k < l && nums[k] == nums[k - 1]);
                    do l--; while (k < l && nums[l] == nums[l + 1]);
                } else if (num < target) {
                    do k++; while (k < l && nums[k] == nums[k - 1]);
                } else {
                    do l--; while (k < l && nums[l] == nums[l + 1]);
                }
            }
        }
    }
    return res;
}


























